### PR TITLE
ceph: stop using radosgw-admin CLI for s3 user management

### PR DESCRIPTION
**Pull request by leseb**
We have been having many issues with external mode with Ceph version
mismatching. The operator would have a Ceph version different than the
external cluster. The `radosgw-admin` was used to interact with S3
users, even a small version delta would cause the command to coredump.
After checking with the rgw core team it appears Rook was misusing the
CLI and the admin ops API should be used instead.
So this patch is the first introduction of go-ceph in Rook to consume
the rgw admin ops API instead of the `radosgw-admin` CLI, **only** for
user management in this initial commit.
Later we can do more such as bucket operation, zone management etc.

Closes: https://github.com/rook/rook/issues/7924
Signed-off-by: Sébastien Han <seb@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

Depends on https://github.com/ceph/go-ceph/pull/496

### CODE REVIEW

File: pkg/operator/ceph/object/admin.go
Comment (travisn) [inline pkg/operator/ceph/object/admin.go:209]: Instead of storing the admin ops user creds in a secret, can we just query them from Ceph when needed if it's a local ceph cluster? We are anyways querying them from ceph here, right?

Comment (leseb) [reply pkg/operator/ceph/object/admin.go:209]: Yes we are. Sounds good to me.

**Comment by thotz** [comment]
@leseb : I have a general question, how will the `user` needed for s3 admin ops created? Do rook creates it via `radosgw-admin` commands or user need to create it manually after setting up ceph object store?

**Comment by leseb** [comment]
> @leseb : I have a general question, how will the `user` needed for s3 admin ops created? Do rook creates it via `radosgw-admin` commands or user need to create it manually after setting up ceph object store?

Yes we still create that user once with `radosgw-admin` and never come back to it. It's safe in the converged scenario since we rarely hit incompatibility issues (never hit any actually). For the external mode case, we create the user on the external cluster and export the credentials as part of the son output.

File: pkg/operator/ceph/object/health.go
Comment (thotz) [inline pkg/operator/ceph/object/health.go:132]: @leseb: Is there any specific reason to replace `admin ops user` with `health checker`. Both serve different use cases right. IMO it will help to keep them separate if we want to debug some issues with respect to each feature

File: pkg/operator/ceph/object/user.go
Comment (thotz) [inline pkg/operator/ceph/object/user.go:153]: @leseb : Why this api is removed? I agree currently this API is not used anywhere. But in future we might need this API. There was a request from an upstream user to support quota for `cephobjectstoreuser`. In those cases this API will helpful

Comment (leseb) [reply pkg/operator/ceph/object/user.go:153]: I was thinking we could use `ModifyUser()` from the API when the time comes. There is no need for `radosgw-admin` commands anymore.

File: pkg/operator/ceph/object/health.go
Comment (leseb) [reply pkg/operator/ceph/object/health.go:132]: Well, the goal the healtchecher is to ensure the object store works. So I don't see any problem with using the admin ops user to validate this. Why is it a concern?

Comment (thotz) [reply pkg/operator/ceph/object/health.go:132]: @leseb: It is purely from debugging point of view.
`healthchecker` creates buckets and put data into a bucket frequently.  But `admin ops` user will handle requests for OBC, cephobjectstoreuser etc.

File: pkg/operator/ceph/object/user.go
Comment (thotz) [reply pkg/operator/ceph/object/user.go:153]: Okay I agree. Please don't forget to add it back

File: pkg/operator/ceph/object/health.go
Comment (leseb) [reply pkg/operator/ceph/object/health.go:132]: Hum so you want to create a checker user with the admin API and then use it?

Comment (leseb) [reply pkg/operator/ceph/object/health.go:132]: Done!

File: pkg/operator/ceph/object/bucket/provisioner.go
Comment (thotz) [inline pkg/operator/ceph/object/bucket/provisioner.go:354]: @leseb : Just wondering what this means for each request on Grant() or Provision() will result from the creation of adminsOpsClient() do we need really need that, instead can we use the same `adminopsclient` initialised as part of cephobjectstore and use that refer here.

Secondly `p.accessKeyID`, `p.secretAccesskey` specific to user of each OBC request which does not have capabilities to remove or create user. This is used to pass back to OBC request. For Provision() call, we may need to create bucket with this credentials otherwise we just give back details in OB response

File: pkg/operator/ceph/object/controller.go
Comment (travisn) [inline pkg/operator/ceph/object/controller.go:85]: Can the key and secret be on the ReconcileCephObjectStore struct? They are different per object store, right? I believe this struct is used for reconciling all object stores?

File: pkg/operator/ceph/object/admin.go
Comment (travisn) [inline pkg/operator/ceph/object/admin.go:191]: nit: if you assign it to the local var, you could use it below instead of referencing `s.Data[]` again
```suggestion
		if accessKey, ok := s.Data[rgwAdminOpsUserAccessKey]; !ok {
```

File: pkg/operator/ceph/object/controller.go
Comment (travisn) [inline pkg/operator/ceph/object/controller.go:420]: Is this reconcileRGWAdminOpsUser() the same as above on lines 363-367? Looks like it's the same call for external and local clusters. Could this be factored out in common? In other words, move it out of the if/else block?

File: pkg/operator/ceph/object/health.go
Comment (travisn) [inline pkg/operator/ceph/object/health.go:49]: This object context seems like the right struct where the creds should be stored if possible.

File: pkg/operator/ceph/object/health.go
Comment (travisn) [inline pkg/operator/ceph/object/health.go:170]: Will all error cases return before this? Seems like this else block isn't needed, and also wouldn't need lines 165-166 that seem duplicated above.
```suggestion
	} 
	// Set access and secret key
	s3AccessKey = user.Keys[0].AccessKey
	s3SecretKey = user.Keys[0].SecretKey<br class="Apple-interchange-newline">
```

File: pkg/operator/ceph/object/health.go
Comment (travisn) [inline pkg/operator/ceph/object/health.go:159]: Do CreateUser() and GetUser() return the same type? If so, we can re-use the var to have common assign code later?
```suggestion
			user, err := co.GetUser(context.Background(), userConfig)
```

**Review by travisn — CHANGES_REQUESTED** [review]
Any ideas on the object test failures in the smoke and upgrade suites?

File: pkg/operator/ceph/object/bucket/provisioner.go
Comment (leseb) [reply pkg/operator/ceph/object/bucket/provisioner.go:354]: > @leseb : Just wondering what this means for each request on Grant() or Provision() will result from the creation of adminsOpsClient() do we need really need that? instead can we use the same `adminopsclient` initialised as part of cephobjectstore and use that refer here.


How?

> Secondly `p.accessKeyID`, `p.secretAccesskey` specific to user of each OBC request which does not have capabilities to remove or create the user. This is used to pass back to OBC request. For Provision() call, we may need to create a bucket with these credentials otherwise we just give back details in OB response and delete the bucket in Delete()

Good point, we should use we admin user instead. Will look deeper.

File: pkg/operator/ceph/object/controller.go
Comment (leseb) [reply pkg/operator/ceph/object/controller.go:420]: I can try.

File: pkg/operator/ceph/object/health.go
Comment (leseb) [reply pkg/operator/ceph/object/health.go:49]: Will try.

File: pkg/operator/ceph/object/bucket/provisioner.go
Comment (thotz) [reply pkg/operator/ceph/object/bucket/provisioner.go]: @leseb: My first question was there is `adminOpsClient        *admin.API` part of `context` of `cephobjectstore`. This getting initialised in `checkObjectStoreHealth()`. Instead of doing that, can we  intialise that on `reconcileCreateObjectStore()` and use that `adminOpsClient` from `cephobjectstorecontext` for all `OBC` and `cephobjectstoreusers`. You might need to intialise the `adminOpsClient` before  `checkObjectStoreHealth()` for addressing `dashboard` usecase.

File: pkg/operator/ceph/object/admin.go
Comment (thotz) [inline pkg/operator/ceph/object/admin.go:198]: @leseb : IMO it is better to add the `capabilities` for `AdminOpsUser` than setting it into `SystemUser`. AFAIK `SystemUser` is used to address special cases like `dashboard` or `mutlisite-sync` and not exposed for external S3 use. By setting the `SystemUser` we might end up having certain limitations

**Review by thotz — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/object/admin.go
Comment (leseb) [reply pkg/operator/ceph/object/admin.go:198]: What about using `admin` then?

Comment (thotz) [reply pkg/operator/ceph/object/admin.go:198]: Sorry, I am not sure about the capabilities of `--admin` flag in the user. The [doc](https://docs.ceph.com/en/latest/radosgw/admin/#add-remove-admin-capabilities) mentions adding admin caps via `caps` options for a user

**Comment by leseb** [comment]
I need https://github.com/ceph/go-ceph/pull/512 to complete the implementation.

File: pkg/operator/ceph/object/admin.go
Comment (leseb) [reply pkg/operator/ceph/object/admin.go:198]: I think the doc is outdated but the `--admin` seems pretty clear to me that it will enable admin caps to a given user. That's probably a way to avoid passing all the caps.

Comment (thotz) [reply pkg/operator/ceph/object/admin.go:198]: Okay if it working as expected then I am not against using it

File: pkg/operator/ceph/object/bucket/provisioner.go
Comment (travisn) [inline pkg/operator/ceph/object/bucket/provisioner.go:237]: can be removed?

File: pkg/operator/ceph/object/bucket/provisioner.go
Comment (travisn) [inline pkg/operator/ceph/object/bucket/provisioner.go:397]: This block looks the same as above, can it be factored to a helper method? Or maybe `setAdminOpsAPIClient()` just needs to do the special error handling?

Comment (leseb) [reply pkg/operator/ceph/object/bucket/provisioner.go:397]: I'm not sure there is much to factor in here, saving one line, it's already the top-level caller too. Also, yes they are the same just like a bunch of code since both are called under different circomstances.

File: pkg/operator/ceph/object/bucket/provisioner.go
Comment (thotz) [inline pkg/operator/ceph/object/bucket/provisioner.go:140]: nit: above commands says `-1` represents restricting of creating new buckets, but variable `unlimitedBuckets` is bit confusing here

File: pkg/operator/ceph/object/bucket/provisioner.go
Comment (thotz) [inline pkg/operator/ceph/object/bucket/provisioner.go:246]: I fully don't understand this check, is it ignoring `No such user error`?

**Review by thotz — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/object/bucket/provisioner.go
Comment (leseb) [reply pkg/operator/ceph/object/bucket/provisioner.go:574]: 😱

File: pkg/operator/ceph/object/bucket/provisioner.go
Comment (leseb) [reply pkg/operator/ceph/object/bucket/provisioner.go:585]: I just moved up the quota enablement and removed the `Enabled: &quotaEnabled` from the conditional calls.

File: pkg/operator/ceph/object/bucket/provisioner.go
Comment (leseb) [reply pkg/operator/ceph/object/bucket/provisioner.go:140]: Oh I thought `-1` meant unlimited buckets. Thanks

File: pkg/operator/ceph/object/bucket/provisioner.go
Comment (leseb) [reply pkg/operator/ceph/object/bucket/provisioner.go:246]: I'll change that

File: pkg/operator/ceph/object/user.go
Comment (leseb) [reply pkg/operator/ceph/object/user.go:47]: Most of them are not used, I just going to remove them.

File: pkg/operator/ceph/object/admin.go
Comment (leseb) [reply pkg/operator/ceph/object/admin.go]: I'm gonna use `caps` so we have a clear understanding of the permissions for that user.

**Comment by travisn** [comment]
Overall LGTM, will approve after you've confirmed your testing and final changes are done

File: pkg/operator/ceph/object/user.go
Comment (thotz) [inline pkg/operator/ceph/object/user.go:46]: Can we pass the `Caps` as a `string` in general instead of `bool` specific to adminOpsUser, so that if `cephObjectStoreUser` extend to support `caps` in future we can avoid adding to the new variable `rgwUserInfo` struct.

File: cluster/examples/kubernetes/ceph/create-external-cluster-resources.sh
Comment (thotz) [inline cluster/examples/kubernetes/ceph/create-external-cluster-resources.sh:82]:
```suggestion
  createRGWAdminOpsUserKeys=$(radosgw-admin user create --uid rgw-admin-ops-user --display-name "Rook RGW Admin Ops user" --caps "buckets=*;users=*;usage=read;metadata=read;zone=read"|jq --raw-output .keys[0])
```

**Review by thotz — COMMENTED** [review]
Overall LGTM. Have two minor comments

File: pkg/operator/ceph/object/user.go
Comment (leseb) [reply pkg/operator/ceph/object/user.go:46]: `rgwUserInfo` is gone but it's a good point for `cephObjectStoreUser` to support caps eventually. This struct type `ObjectUser` is only used when calling the `radosgw-admin` which `cephObjectStoreUser` has stopped using with this patch. If caps are implemented they will go in the spec and be reflected as a `admin.User` type. So I'm not sure there is a concern here.

THoughts?

**Comment by leseb** [comment]
Successfully tested and validated upgrades. @travisn @thotz  I'm done with all my changes. This is ready for a final review.

**Review by thotz — APPROVED** [review]
LGTM

File: pkg/operator/ceph/object/objectstore.go
Comment (travisn) [inline pkg/operator/ceph/object/objectstore.go:756]: is this needed before merge?

File: pkg/operator/ceph/object/user/controller.go
Comment (travisn) [inline pkg/operator/ceph/object/user/controller.go:384]: Could there ever be a userConfig with no keys, which would cause this to crash? Why is `Keys` a slice? Looks like we always expect one key pair.

Comment (leseb) [reply pkg/operator/ceph/object/user/controller.go:384]: I don't think so, we would get an error before when fetching the user. The `Keys` is a slice because a user can have multiple keys, that's how both rgw admin ops API and `radosgw-admin` return the user definition.

File: pkg/operator/ceph/object/objectstore.go
Comment (leseb) [reply pkg/operator/ceph/object/objectstore.go:756]: It's for "future" reference, once we complete the migration.

**Review by travisn — APPROVED** [review]
Ok let's do this!

**Pull request merged**
